src/transformations/defmodel.hpp

```diff
@@ -0,0 +1,1890 @@
+                                double /* b */, double /*es*/, double & /* X */,
+                                double & /* Y */, double & /* Z */) {
+        throw UnimplementedException("geographicToGeocentric unimplemented");
+    }
+
+    void geocentricToGeographic(double /* X */, double /* Y */, double /* Z */,
+                                double /* a */, double /* b */, double /*es*/,
+                                double & /* lam */, double & /* phi */,
+                                double & /* height*/) {
+        throw UnimplementedException("geocentricToGeographic unimplemented");
+    }
+
+    bool isGeographicCRS(const std::string & /* crsDef */) {
+        throw UnimplementedException("isGeographicCRS unimplemented");
+    }
+
+#ifdef DEBUG_DEFMODEL
+    void log(const std::string & /* msg */) {
+        throw UnimplementedException("log unimplemented");
+    }
+#endif
+};
+
+// ---------------------------------------------------------------------------
+
+constexpr double DEFMODEL_PI = 3.14159265358979323846;
+constexpr double DEG_TO_RAD_CONSTANT = 3.14159265358979323846 / 180.;
+inline constexpr double DegToRad(double d) { return d * DEG_TO_RAD_CONSTANT; }
+
+// ---------------------------------------------------------------------------
+
+/** Internal class to offer caching services over a Grid */
+template <class Grid> struct GridEx {
+    const Grid *grid;
+
+    bool smallResx; // lesser than one degree
+
+    double sinhalfresx;
+    double coshalfresx;
+
+    double sinresy;
+    double cosresy;
+
+    int last_ix0 = -1;
+    int last_iy0 = -1;
+    double dX00 = 0;
+    double dY00 = 0;
+    double dZ00 = 0;
+    double dX01 = 0;
+    double dY01 = 0;
+    double dZ01 = 0;
+    double dX10 = 0;
+    double dY10 = 0;
+    double dZ10 = 0;
+    double dX11 = 0;
+    double dY11 = 0;
+    double dZ11 = 0;
+    double sinphi0 = 0;
+    double cosphi0 = 0;
+    double sinphi1 = 0;
+    double cosphi1 = 0;
+
+    explicit GridEx(const Grid *gridIn)
+        : grid(gridIn), smallResx(grid->resx < DegToRad(1)),
+          sinhalfresx(sin(grid->resx / 2)), coshalfresx(cos(grid->resx / 2)),
+          sinresy(sin(grid->resy)), cosresy(cos(grid->resy)) {}
+
+    // Return geocentric offset (dX, dY, dZ) relative to a point
+    // where x0 = -resx / 2
+    inline void getBilinearGeocentric(int ix0, int iy0, double de00,
+                                      double dn00, double de01, double dn01,
+                                      double de10, double dn10, double de11,
+                                      double dn11, double m00, double m01,
+                                      double m10, double m11, double &dX,
+                                      double &dY, double &dZ) {
+        // If interpolating in the same cell as before, then we can skip
+        // the recomputation of dXij, dYij and dZij
+        if (ix0 != last_ix0 || iy0 != last_iy0) {
```

Comment:
Lots of math here and not lots of comment...

Reply:
the maths here come from @ccrok proposal. I suspect they are "traditional" to geodesy foks. I've added a small comment regarding that